File: data_structures/stack/java/StackUsingArray.java
public class StackEmptyException extends Exception {

	/**
	 * 
	 */
	private static final long serialVersionUID = 1L;

}


public class StackUsingArray {
	
	private int data[];
	private int top; // is the index of topmost element of stack
	
	public StackUsingArray() {
		data = new int[10];
		top = -1;
	}
	
	public StackUsingArray(int capacity) {
		data = new int[capacity];
		top = -1;
	}
	
	public  boolean isEmpty(){
//		if(top == -1){
//			return true;
//		}else{
//			return false;
//		}
		return (top == -1);
	}
	
	public int size(){
		return top + 1;	
		}
	
	public int top() throws StackEmptyException{
		if(size() == 0){
			//StackEmptyException
			StackEmptyException e = new StackEmptyException();
			throw e;
		}
		return data[top];
	}
	
	public void push(int elem) //throws StackFullException
	{
		if(size() == data.length){
			// Stack Full
//			StackFullException e = new StackFullException();
//			throw e;	
			int temp[] = data;
			data = new int[temp.length*2];
			for(int i=0;i<temp.length;i++) {
				data[i] = temp[i];
			}
		}
		top++;
		data[top] = elem;
	}

	public int pop() throws StackEmptyException{
		if(size() == 0){
			//StackEmptyException
			StackEmptyException e = new StackEmptyException();
			throw e;
		}
		int temp = data[top];
		top--;
		return temp;
		
	}
}

File: data_structures/stack/cpp/StackUsingArray.cpp
#include<iostream>
using namespace std;
struct stack
{
    int size;
    int top;
    int *s;
};
void create(struct stack *st)
{
    cout<<"enter stack size";
    cin>>st->size;
    st->top=-1;
    st->s=new int[st->size];
}
void display(struct stack *st)
{
    for(int i=st->top;i>=0;i--)
        cout<<st->s[i]<<" ";
}
void push(struct stack *st,int x)
{
    if(st->top==st->size-1)
        cout<<"stack is full";
    else
    {
        st->top++;
        st->s[st->top]=x;
    }
}
int pop(struct stack *st)
{
    int x=-1;
    if(st->top==-1)
        cout<<"stack is underflow";
    else
    {
        x=st->s[st->top];
        st->top--;
    }
    return x;
}
int peek(struct stack *st,int pos)
{
    int x=-1;
    if(st->top-pos+1<0)
        cout<<"invalid index";
    else
    {
        x=st->s[st->top-pos+1];
    }
    return x;
}
int isempty(struct stack *st)
{
    if(st->top==-1)
        return 1;
    else
        return 0;
}
int isfull(struct stack *st)
{
    if(st->top==st->size-1)
        return 1;
    else
        return 0;
}
int stacktop(struct stack *st)
{
    if(!isempty(st))
        return st->s[st->top];
    return 0;
}
int main()
{
    struct stack st;
    create(&st);
    push(&st,3);
    push(&st,5);
    push(&st,6);
    push(&st,7);
    push(&st,9);
    display(&st);
    cout<<endl;
    cout<<pop(&st)<<endl;
    display(&st);
    cout<<endl;
    cout<<pop(&st)<<endl;
    display(&st);
    cout<<endl;
    cout<<pop(&st)<<endl;
    display(&st);
    cout<<endl;
    cout<<pop(&st)<<endl;
    display(&st);
    cout<<endl;
    cout<<pop(&st)<<endl;
    display(&st);
    cout<<endl;
    cout<<pop(&st)<<endl;
    display(&st);
    push(&st,3);
    push(&st,5);
    push(&st,6);
    push(&st,7);
    push(&st,9);
    display(&st);
    cout<<endl;
    cout<< peek(&st,2)<<endl;
    cout<<isempty(&st)<<endl;
    cout<<isfull(&st)<<endl;
    cout<<stacktop(&st);
}
